pkg/migrate/cstor/blockdevice_correction.go

```diff
@@ -0,0 +1,364 @@
+	}
+	// For the new bd
+	// - if bdc already present:
+	//   - if ownerref already present: update the bdc with correct ownerref & label
+	//   - if ownerref not present: add the spc ownerref & label bdc
+
+	if newBDObj.Spec.ClaimRef != nil {
+		newBDCObj, err := c.OpenebsClientset.OpenebsV1alpha1().
+			BlockDeviceClaims(c.OpenebsNamespace).
+			Get(newBDObj.Spec.ClaimRef.Name, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+		if newBDCObj.Labels["openebs.io/storage-pool-claim"] != spcObj.Name {
+			newBDCObj.OwnerReferences = []metav1.OwnerReference{
+				*metav1.NewControllerRef(spcObj,
+					apis.SchemeGroupVersion.WithKind(spcKind)),
+			}
+			if newBDCObj.Labels == nil {
+				newBDCObj.Labels = map[string]string{}
+			}
+			newBDCObj.Labels["openebs.io/storage-pool-claim"] = spcObj.Name
+			newBDCObj, err = c.OpenebsClientset.OpenebsV1alpha1().
+				BlockDeviceClaims(c.OpenebsNamespace).Update(newBDCObj)
+		}
+	} else {
+		// - if bdc not present: remove legacy annotation on bd
+		// this will enable claiming of the bd
+		delete(newBDObj.Annotations, "internal.openebs.io/uuid-scheme")
+		newBDObj, err = c.OpenebsClientset.OpenebsV1alpha1().
+			BlockDevices(c.OpenebsNamespace).Update(newBDObj)
+		if err != nil {
+			return err
+		}
+		resourceList, err := getCapacity(byteCount(newBDObj.Spec.Capacity.Storage))
+		if err != nil {
+			return errors.Errorf("failed to get capacity from block device %s:%s",
+				newBDObj.Name, err)
+		}
+		// claim the bd immediately so no other resource can take it
+		newBDCObj := v1alpha1.NewBlockDeviceClaim().
+			WithName("bdc-cstor-" + string(newBDObj.UID)).
+			WithNamespace(c.OpenebsNamespace).
+			WithLabels(map[string]string{string(apis.StoragePoolClaimCPK): spcObj.Name}).
+			WithBlockDeviceName(newBDObj.Name).
+			WithHostName(newBDObj.Labels[openebstypes.HostNameLabelKey]).
+			WithCSPCOwnerReference(*metav1.NewControllerRef(spcObj,
+				apis.SchemeGroupVersion.WithKind(spcKind))).
+			WithCapacity(resourceList).
+			WithFinalizer("storagepoolclaim.openebs.io/finalizer")
+
+		newBDCObj, err = c.OpenebsClientset.OpenebsV1alpha1().
+			BlockDeviceClaims(c.OpenebsNamespace).Create(newBDCObj)
+
+	retryBDCStatus:
+		newBDCObj, err = c.OpenebsClientset.OpenebsV1alpha1().
+			BlockDeviceClaims(c.OpenebsNamespace).Get(newBDCObj.Name, metav1.GetOptions{})
+		if newBDCObj.Status.Phase != v1alpha1.BlockDeviceClaimStatusDone {
+			klog.Infof("waiting for bdc %s to get bound", newBDObj.Name)
+			time.Sleep(2 * time.Second)
+			goto retryBDCStatus
+		}
+	}
+	// For the old bd
+	// - if active: remove the ownerref & label from corresponding bd claim present
+	// - if inactive: leave it as it is
+	if oldBDObj.Status.State == "Active" {
+		oldBDCObj, err := c.OpenebsClientset.OpenebsV1alpha1().
+			BlockDeviceClaims(c.OpenebsNamespace).
+			Get(oldBDObj.Spec.ClaimRef.Name, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+		if oldBDCObj.Labels["openebs.io/storage-pool-claim"] == spcObj.Name {
```

Comment:
```suggestion
		if oldBDCObj.Labels["openebs.io/storage-pool-claim"] != spcObj.Name {
```

Reply:
Here we are removing the ownerref & label from the bdc which no longer belongs to the pool. So we should only remove if it belongs to that pool. Right?

Reply:
If old bd belongs to the same CSP(SPC) why we need to delete it?

Reply:
So the map is like
oldBD -- got converted to --> newBD
The flow should be

Reply:
We have to remove only if ownerRef and labels belong to a different pool

Reply:
>remove ownerref & label from oldBD if it belong to same csp because it might belong to some other csp now.
Right?

Ok got it I made miss assumption... Current check is good

Reply:
Anyways oldBD will always belong to the current CSP/SPC